lib/ethers/multicall.ex

```diff
@@ -0,0 +1,164 @@
+defmodule Ethers.Multicall do
+  @moduledoc """
+  Multicall token interface
+
+  More info: https://www.multicall3.com
+  """
+
+  import Ethers.Utils, only: [hex_decode!: 1]
+
+  alias Ethers.Contracts.Multicall3
+  alias Ethers.ExecutionError
+  alias Ethers.TxData
+
+  def aggregate3(data) when is_list(data) do
+    data
+    |> Enum.map(&aggregate3_encode_data/1)
+    |> Multicall3.aggregate3()
+  end
+
+  @spec aggregate3_encode_data({Ethers.TxData.t()} | {Ethers.TxData.t(), keyword()}) ::
+          {any(), boolean(), binary()}
+  def aggregate3_encode_data(data)
+
+  def aggregate3_encode_data({%TxData{data: data}, opts}) do
+    {Keyword.fetch!(opts, :to), Keyword.get(opts, :allow_failure, true), hex_decode!(data)}
+  end
+
+  def aggregate3_encode_data({%TxData{data: data, default_address: address}, opts})
+      when not is_nil(address) do
+    {address, Keyword.get(opts, :allow_failure, true), hex_decode!(data)}
+  end
+
+  def aggregate3_encode_data({%TxData{data: data, default_address: address}})
+      when not is_nil(address) do
+    {address, true, hex_decode!(data)}
+  end
+
+  def aggregate3_encode_data(%TxData{data: data, default_address: address})
+      when not is_nil(address) do
+    {address, true, hex_decode!(data)}
+  end
+
+  def aggregate2(data) when is_list(data) do
+    data
+    |> Enum.map(&aggregate2_encode_data/1)
+    |> Multicall3.aggregate()
+  end
+
+  def aggregate2_encode_data(data)
+
+  def aggregate2_encode_data({%TxData{data: data}, opts}) do
+    {Keyword.fetch!(opts, :to), hex_decode!(data)}
+  end
+
+  def aggregate2_encode_data({%TxData{data: data, default_address: address}})
+      when not is_nil(address) do
+    {address, hex_decode!(data)}
+  end
+
+  def aggregate2_encode_data(%TxData{data: data, default_address: address})
+      when not is_nil(address) do
+    {address, hex_decode!(data)}
+  end
+
+  @doc """
+  Makes an eth_call to with the given data and overrides using Ethers.call/2.
+  The responses are parsed and decoded using `:calls` provided in the overrides.
+  Assumes the call is to Multicall, either from a prepared aggregate2 or aggregate3.
+
+  ## Overrides and Options
+    All from `Ethers.call/2`, and:
+
+    - `:calls`: List of function calls or selectors. Used for decoding results.
+
+  ## Examples
+
+  ```elixir
+  calls = [ ContractA.foo(), { ContractB.foo(), to: "0x..." } ]
+  calls |> Ethers.Multicall.aggregate3() |> Ethers.Multicall.call(calls: calls)
```

Comment:
Why would we need `calls: calls` in the end?

Reply:
To preserve API compatibility with `Ethers.call/2`. If we piped `calls` into `Ethers.call/2` it wouldnt work, however it does mean using `Ethers.Multicall.call/2` is a bit ugly